pkg/container/qemu_runner.go

```diff
@@ -0,0 +1,820 @@
+		0644)
+	if err != nil {
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: setup DNS")
+	err = os.WriteFile(filepath.Join(rootfs, "etc/resolv.conf"),
+		[]byte("nameserver 1.1.1.1"),
+		0644)
+	if err != nil {
+		return "", "", err
+	}
+
+	// allow passing env variables to ssh commands
+	clog.FromContext(ctx).Debug("qemu: ssh - acceptenv")
+	sshdConfig, err := os.OpenFile(filepath.Join(rootfs, "etc/ssh/sshd_config"), os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		return "", "", err
+	}
+	_, err = sshdConfig.WriteString(`AcceptEnv *`)
+	if err != nil {
+		return "", "", err
+	}
+	defer sshdConfig.Close()
+
+	clog.FromContext(ctx).Debug("qemu: ssh - create ssh key pair")
+	pubKeyBytes, err := generateSSHKeys(ctx, cfg, rootfs)
+	if err != nil {
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: ssh - inject authorized_keys - root")
+	err = os.WriteFile(filepath.Join(rootfs, "root/.ssh/authorized_keys"), pubKeyBytes, 0400)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: ssh pubkey write failed: %v", err)
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: ssh - fix dir permissions - root")
+	err = os.Chmod(filepath.Join(rootfs, "root/.ssh"), 0700)
+	if err != nil {
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: ssh - inject authorized_keys - build user")
+	err = os.WriteFile(filepath.Join(rootfs, "home/build/.ssh/authorized_keys"), pubKeyBytes, 0400)
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: ssh pubkey write failed: %v", err)
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: ssh - fix dir permissions - build user")
+	err = os.Chmod(filepath.Join(rootfs, "home/build/.ssh"), 0700)
+	if err != nil {
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: setting up kernel for vm")
+	kernel, err := generateVmlinuz(ctx, rootfs)
+	if err != nil {
+		return "", "", err
+	}
+
+	clog.FromContext(ctx).Debug("qemu: generating initramfs...")
+	initramfs, err := generateInitrd(ctx, rootfs)
+	if err != nil {
+		return "", "", err
+	}
+
+	return kernel, initramfs, nil
+}
+
+func generateVmlinuz(ctx context.Context, rootfs string) (string, error) {
+	clog.FromContext(ctx).Info("qemu: detecting kernel")
+	if _, err := os.Stat("/boot/vmlinuz"); err == nil {
+		clog.FromContext(ctx).Info("qemu: vmlinuz detected, reusing it")
+		return "/boot/vmlinuz", nil
+	}
+
+	clog.FromContext(ctx).Info("qemu: detecting kernel")
+	if _, err := os.Stat("/boot/vmlinuz-virt"); err == nil {
+		clog.FromContext(ctx).Info("qemu: vmlinuz detected, reusing it")
+		return "/boot/vmlinuz-virt", nil
+	}
+
+	clog.FromContext(ctx).Info("qemu: kernel not found, downloading...")
+	// download mainline kernel from ubuntu
+	response, err := http.Get("https://dl-cdn.alpinelinux.org/alpine/edge/main/x86_64/linux-virt-6.6.41-r0.apk")
+	if err != nil {
+		clog.FromContext(ctx).Errorf("qemu: can't download kernel - %v", err)
+		return "", err
+	}
+	defer response.Body.Close()
```

Comment:
Instead of hardcoding this (which will break when Alpine withdraws this kernel, and won't work on aarch64) perhaps we can allow folks to set an environment variable for where to pick up the kernel (if set)?

Reply:
Makes sense yes, implemented it so that one can pass a file path to a prebuild vmlinuz or an APK url
else it will use a sensible default